jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,559 @@
+     */
+    public DosHandler(
+        @Name("handler") Handler handler,
+        @Name("getId") Function<Request, String> getId,
+        @Name("rateFactory") RateControl.Factory rateControlFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        super(handler);
+        installBean(_trackers);
+        _getId = Objects.requireNonNullElse(getId, ID_FROM_REMOTE_ADDRESS);
+        installBean(_getId);
+        _rateControlFactory = Objects.requireNonNull(rateControlFactory);
+        installBean(_rateControlFactory);
+        _maxTrackers = maxTrackers <= 0 ? 10_000 : maxTrackers;
+        _rejectHandler = Objects.requireNonNullElseGet(rejectHandler, StatusRejectHandler::new);
+        installBean(_rejectHandler);
+    }
+
+    @Override
+    public void setServer(Server server)
+    {
+        super.setServer(server);
+        if (_rejectHandler instanceof Handler handler)
+            handler.setServer(server);
+    }
+
+    @Override
+    protected boolean onConditionsMet(Request request, Response response, Callback callback) throws Exception
+    {
+        // Reject if we have too many Trackers
+        if (_maxTrackers > 0 && _trackers.size() >= _maxTrackers)
+        {
+            // Try shrinking the tracker pool
+            long now = NanoTime.now();
+            _trackers.values().removeIf(tracker -> tracker.isIdle(now));
+            if (_trackers.size() >= _maxTrackers)
+                return _rejectHandler.handle(request, response, callback);
+        }
+
+        // Calculate an id for the request (which may be global empty string)
+        String id;
+        id = _getId.apply(request);
+
+        if (id == null)
+            return _rejectHandler.handle(request, response, callback);
+
+        // Obtain a tracker, creating a new one if necessary.  Trackers are removed if CyclicTimeouts#onExpired returns true
+        Tracker tracker = _trackers.computeIfAbsent(id, this::newTracker);
+
+        // If we are not over-limit then handle normally
+        if (!tracker.isRateExceeded(request.getBeginNanoTime(), true))
+            return nextHandler(request, response, callback);
+
+        // Otherwise reject the request
+        return _rejectHandler.handle(request, response, callback);
+    }
+
+    Tracker newTracker(String id)
+    {
+        return new Tracker(id, _rateControlFactory.newRate());
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        _cyclicTimeouts = new CyclicTimeouts<>(getServer().getScheduler())
+        {
+            @Override
+            protected Iterator<Tracker> iterator()
+            {
+                return _trackers.values().iterator();
+            }
+
+            @Override
+            protected boolean onExpired(Tracker tracker)
+            {
+                return tracker.isIdle(NanoTime.now());
```

Comment:
I don't think this is right.
This method is called when the tracker is expired already. There is no point asking whether the tracker is idle, because we know it is.

Reply:
No, the tracker is expired one idleCheck period after the last sample.  Depending on how the tracker is configured, that may or may not be truly idle.  Depending on the actual Tracker and for the standard Tracker, how the alpha is configured, the rate may decrease slowly and the tracker may not be idle after one check period.

Reply:
I don't understand.

The idleness is either determined by time, and/or by the specific rate control algorithm.

For example, with a moving window, if you are idle for as long as the sample period, you have no more samples, so you are idle.
The implementation can just set the `_expireAt` are the right time, so there is no need to have a method to ask.

For another algorithm, it would know differently when it is idle, and therefore set `_expireAt` at the right time.

So perhaps we need to have the `Tracker` delegate to the `RateControl`, either always (`RateControl extends Expirable`) or optionally by having `Tracker` test `RateControl instanceof Expirable` (otherwise defaulting).

Reply:
@sbordet I don't think you are understanding how sliding averages work.  It is simply not correct when you say "if you are idle for as long as the sample period, you have no more samples, so you are idle". 

If the current rate is measured at 100r/s for a check period of 1s, then after 1s with no requests arriving the measured rate will depend on the alpha setting, but it could be anything from 0 r/s to 50r/s